### PR TITLE
[TOPIC-BLE-LLCP] Bluetooth: controller: Adding Central/Peripheral compile dependencies to new LLCP 

### PR DESCRIPTION
**Pull request by erbr-ot**
Compiling only the code needed depending on CONFIG_BT_CENTRAL/CONFIG_BT_PERIPHERAL for procedures:
Feature Exchange, Min Used Channels, PHY Update and Encryption

Remaining procedures will be visited when complete.

Compile permutations script updated to compile all permutations so far, except ones with CENTRAL==PERIPHERAL==n

### CODE REVIEW

File: subsys/bluetooth/controller/hci/hci.c [hunk 1]
```diff
@@ -2190,6 +2191,7 @@ static void le_read_remote_features(struct net_buf *buf, struct net_buf **evt)
 
 	*evt = cmd_status(status);
 }
+#endif /* CONFIG_BT_CENTRAL) || CONFIG_BT_CTLR_SLAVE_FEAT_REQ */
```
Comment (tosk-ot) [inline subsys/bluetooth/controller/hci/hci.c:2194]: CONFIG_BT_CENTRAL) - there is a ")" too much - also in endifs below :o)

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 2]
```diff
@@ -234,22 +242,28 @@ struct proc_ctx *ull_cp_priv_create_remote_procedure(enum llcp_proc proc)
...
+#endif /* CONFIG_BT_CTLR_MIN_USED_CHAN */
 	case PROC_VERSION_EXCHANGE:
 		rp_comm_init_proc(ctx);
 		break;
+#if defined (CONFIG_BT_CTLR_LE_ENC) && defined (CONFIG_BT_PERIPHERAL)
 	case PROC_ENCRYPTION_START:
 	case PROC_ENCRYPTION_PAUSE:
 		rp_enc_init_proc(ctx);
 		break;
+#endif /* CONFIG_BT_CTLR_LE_ENC */
```
Comment (tosk-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:266]: missing CONFIG_BT_PERIPHERAL in comment?

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 3]
```diff
@@ -395,6 +419,7 @@ uint8_t ull_cp_feature_exchange(struct ll_conn *conn)
 
 	return BT_HCI_ERR_SUCCESS;
 }
+#endif /* CONFIG_BT_CENTRAL) || CONFIG_BT_CTLR_SLAVE_FEAT_REQ */
```
Comment (tosk-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:422]: Woops :o) you're too generous with ")"s in comment

File: subsys/bluetooth/controller/ll_sw/ull_llcp_hci.c [hunk 4]
```diff
@@ -137,8 +140,7 @@ uint8_t ll_feature_req_send(uint16_t handle)
 
 	return ull_cp_feature_exchange(conn);
 }
-
-
+#endif /* CONFIG_BT_CENTRAL) || CONFIG_BT_CTLR_SLAVE_FEAT_REQ */
```
Comment (tosk-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_hci.c:143]: CONFIG_BT_CENTRAL)

File: subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c [hunk 5]
```diff
@@ -202,26 +205,39 @@ static uint8_t pu_check_update_ind(struct ll_conn *conn, struct proc_ctx *ctx)
...
 static uint8_t pu_apply_phy_update(struct ll_conn *conn, struct proc_ctx *ctx)
 {
 	struct lll_conn *lll = &conn->lll;
 
-	if (ctx->data.pu.s_to_m_phy) {
-		if (lll->role == BT_HCI_ROLE_SLAVE) {
+	if (0) {}
+#if defined(CONFIG_BT_PERIPHERAL)
+	else if (lll->role == BT_HCI_ROLE_SLAVE)
+	{
```
Comment (tosk-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:217]: "{" should go on the same line as "if" in this function.

File: subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c [hunk 6]
```diff
@@ -887,6 +946,7 @@ static void rp_pu_st_wait_tx_ack_phy(struct ll_conn *conn, struct proc_ctx *ctx,
 				rp_pu_complete(conn, ctx, evt, param);
 			}
 		}
+#endif /* CONFIG_BT_CENTRAL */
```
Comment (tosk-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:949]: I think CI will require an else after an else-if, even if empty.
And it must contain a comment (like /* do nothing */ or something).

**Comment by erbr-ot** [comment]
rebased

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c [hunk 7]
```diff
@@ -202,26 +205,36 @@ static uint8_t pu_check_update_ind(struct ll_conn *conn, struct proc_ctx *ctx)
...
 
 static uint8_t pu_apply_phy_update(struct ll_conn *conn, struct proc_ctx *ctx)
 {
 	struct lll_conn *lll = &conn->lll;
 
-	if (ctx->data.pu.s_to_m_phy) {
-		if (lll->role == BT_HCI_ROLE_SLAVE) {
+	if (0) {}
+#if defined(CONFIG_BT_PERIPHERAL)
+	else if (lll->role == BT_HCI_ROLE_SLAVE) {
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:216]: Hmm, well yeah, I was hoping we could change it, so as to not have to have the trailing } after the #if defined()?

File: subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c [hunk 8]
```diff
@@ -862,23 +910,28 @@ static void rp_pu_st_wait_tx_phy_rsp(struct ll_conn *conn, struct proc_ctx *ctx,
...
 
 static void rp_pu_st_wait_tx_ack_phy(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
 {
 	switch (evt) {
 	case RP_PU_EVT_ACK:
-		if (ctx->state == RP_PU_STATE_WAIT_TX_ACK_PHY_RSP) {
-			LL_ASSERT(conn->lll.role == BT_HCI_ROLE_SLAVE);
+		if (0) {}
+#if defined(CONFIG_BT_PERIPHERAL)
+		else if (ctx->state == RP_PU_STATE_WAIT_TX_ACK_PHY_RSP) {
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:921]: As above?

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 3 above)
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: Pokey pokey

File: subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:216]: Think we need to stick to the patterns that we use

File: subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:911]: Yes

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:216]: Okey dokey

File: subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c [hunk 9]
```diff
@@ -488,6 +508,7 @@ static void lp_pu_st_wait_rx_phy_rsp(struct ll_conn *conn, struct proc_ctx *ctx,
 		pu_combine_phys(conn, ctx, tx_pref, rx_pref);
 		lp_pu_send_phy_update_ind(conn, ctx, evt, param);
 		break;
+#endif /* CONFIG_BT_CENTRAL */
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c]: Cleaned up

File: subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c [hunk 10]
```diff
@@ -664,6 +691,7 @@ static void lp_pu_execute_fsm(struct ll_conn *conn, struct proc_ctx *ctx, uint8_
 	case LP_PU_STATE_WAIT_TX_ACK_PHY_REQ:
 		lp_pu_st_wait_tx_ack_phy_req(conn, ctx, evt, param);
 		break;
+#if defined(CONFIG_BT_CENTRAL)
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:684]: Should only be compild in for central as peripheral will not get to rx_rsp state

File: subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:911]: Alright then

File: subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c [hunk 11]
```diff
@@ -862,23 +910,28 @@ static void rp_pu_st_wait_tx_phy_rsp(struct ll_conn *conn, struct proc_ctx *ctx,
...
 static void rp_pu_st_wait_tx_ack_phy(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
 {
 	switch (evt) {
 	case RP_PU_EVT_ACK:
-		if (ctx->state == RP_PU_STATE_WAIT_TX_ACK_PHY_RSP) {
-			LL_ASSERT(conn->lll.role == BT_HCI_ROLE_SLAVE);
+		if (0) {}
+#if defined(CONFIG_BT_PERIPHERAL)
+		else if (ctx->state == RP_PU_STATE_WAIT_TX_ACK_PHY_RSP) {
+		    LL_ASSERT(conn->lll.role == BT_HCI_ROLE_SLAVE);
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c]: yeap

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 12]
```diff
@@ -594,8 +590,12 @@ static void rp_comm_rx_decode(struct ll_conn *conn, struct proc_ctx *ctx, struct
 		/* ping_req has no data */
 		break;
 #endif /* CONFIG_BT_CTLR_LE_PING */
+#if defined(CONFIG_BT_PERIPHERAL)
 	case PDU_DATA_LLCTRL_TYPE_FEATURE_REQ:
+#endif /* CONFIG_BT_PERIPHERAL */
+#if defined(CONFIG_BT_CTLR_SLAVE_FEAT_REQ) && defined (CONFIG_BT_CENTRAL)
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c]: yeap

**Review by thoh-ot — APPROVED** [review]

**Review by kruithofa — APPROVED** [review]

**Pull request merged**